src/main/scala/com/fulcrumgenomics/umi/FilterConsensusReads.scala

```diff
@@ -201,34 +207,87 @@ class FilterConsensusReads
-    logger.info(f"Masked ${maskedBases}%,d of ${totalBases}%,d bases in retained primary consensus reads.")
+    logger.info(f"Output $keptReads%,d of $totalReads%,d primary consensus reads.")
+    logger.info(f"Masked $maskedBases%,d of $totalBases%,d bases in retained primary consensus reads.")
+  }
+
+  /** Builds the writer to which filtered records should be written.
+    *
+    *  If the input order is [[SamOrder.Queryname]] or query grouped, then the filtered records will also be in the same
+    *  order.  So if the output order is specified AND does not match the the input order, sorting will occur.
+    *
+    *  If the input order is not [[SamOrder.Queryname]] or query grouped, then the input records will be resorted into
+    *  [[SamOrder.Queryname]].  So if the output order is specified AND is not [[SamOrder.Queryname]], sorting will occur.
+    *
+    *  Otherwise, we can skip sorting!
+    *
+    * */
+  private def buildOutputWriter(inHeader: SAMFileHeader, refMap: Map[Int, ReferenceSequence]): Writer[SamRecord] with Closeable = {
+    val outHeader    = inHeader.clone()
+
+    val inSortOrder  = inHeader.getSortOrder
+    val inGroupOrder = inHeader.getGroupOrder
+    val inSubSort    = Option(inHeader.getAttribute("SS"))
+
+    // Get the order after filtering
+    val (afterFilteringSortOrder, afterFilteringGroupOrder, afterFilteringSubSort) = {
+      if (inSortOrder == SortOrder.queryname || inGroupOrder == GroupOrder.query) { // no sorting occurred, so same as input
+        (inSortOrder, inGroupOrder, inSubSort)
+      }
+      else { // sorting occurred, so it's queryname
+        val order = SamOrder.Queryname
+        (order.sortOrder, order.groupOrder, order.subSort)
+      }
+    }
+
+    // Get the desired output order
+    val (outputSortOrder, outputGroupOrder, outputSubSort) = this.sortOrder match {
+      case None        => (inSortOrder, inGroupOrder, inSubSort) // same as input
+      case Some(order) => (order.sortOrder, order.groupOrder, order.subSort) // specific output
+    }
+
+    val sort: Option[SamOrder] = {
+      // if the order after filtering and the output order match, no need to re-sort the output
+      if (afterFilteringSortOrder == outputSortOrder && afterFilteringGroupOrder == outputGroupOrder && afterFilteringSubSort == outputSubSort) {
+        None
+      } else { // output order and order after filtering do not match, we need to re-sort the output
+        SamOrder.values.find { order =>
+          order.sortOrder == outputSortOrder && order.groupOrder == outputGroupOrder && order.subSort == outputSubSort
+        }.orElse {
+          // this can only happen if the input order is unrecognized
+          throw new IllegalArgumentException(
+            s"The input BAM had an unrecognized sort order (SO:$inSortOrder GO:$inGroupOrder SS: $inSubSort)" +
+            s"\nTry re-running with --sort-order for a supported output order." in $input
+          )
+        }
+      }
```

Comment:
Can't this whole `else` block now just be `this.sortOrder`?  I think your intent is:
1. If `this.sortOrder` is None, and input is query sorted or grouped, no sorting occurs anywhere and output == input
2. Else if `this.sortOrder` is None, and input is _not_ query sorted or grouped, sorting to queryname occurs, and output is in SamOrder.Queryname
3. Else output will be in `this.sortOrder` and sorting will occur _only_ if the `afterFiltering` ordering is different than `this.sortOrder`.

In this `else` branch we've confirmed that afterFiltering and output orderings are different, and that can _only_ occur if `this.sortOrder` is defined, so we can just use it?